### PR TITLE
add functional embedding

### PR DESCRIPTION
**Pull request by hughperkins**
add functional embedding

### CODE REVIEW

**Comment by hughperkins** [comment]
with new test case:

```
~/git/pytorch$ BACKEND=cpu py.test -v test/test_nn.py -k mbedding
================================================= test session starts ==================================================
platform linux -- Python 3.6.1, pytest-3.1.2, py-1.4.34, pluggy-0.4.0 -- /home/ubuntu/conda3/envs/pytorch-dev/bin/python
cachedir: .cache
rootdir: /home/ubuntu/git/pytorch, inifile:
collected 424 items

test/test_nn.py::TestNN::test_CosineEmbeddingLoss PASSED
test/test_nn.py::TestNN::test_CosineEmbeddingLoss_cuda SKIPPED
test/test_nn.py::TestNN::test_CosineEmbeddingLoss_margin PASSED
test/test_nn.py::TestNN::test_CosineEmbeddingLoss_margin_cuda SKIPPED
test/test_nn.py::TestNN::test_Embedding PASSED
test/test_nn.py::TestNN::test_EmbeddingBag PASSED
test/test_nn.py::TestNN::test_EmbeddingBag_cuda SKIPPED
test/test_nn.py::TestNN::test_Embedding_cuda SKIPPED
test/test_nn.py::TestNN::test_Embedding_sparse PASSED
test/test_nn.py::TestNN::test_Embedding_sparse_cuda SKIPPED
test/test_nn.py::TestNN::test_HingeEmbeddingLoss PASSED
test/test_nn.py::TestNN::test_HingeEmbeddingLoss_cuda SKIPPED
test/test_nn.py::TestNN::test_HingeEmbeddingLoss_margin PASSED
test/test_nn.py::TestNN::test_HingeEmbeddingLoss_margin_cuda SKIPPED
test/test_nn.py::TestNN::test_embedding_functional PASSED
test/test_nn.py::TestNN::test_embedding_padding_idx PASSED
```

**Comment by hughperkins** [comment]
Fixed the lint check, and the weird pasting-into-vim-over-a-slow-connection-does-strange-things comments :-)

File: torch/nn/functional.py [hunk 1]
```diff
@@ -550,6 +550,19 @@ def bilinear(input1, input2, weight, bias=None):
         return Bilinear.apply(input1, input2, weight, bias)
 
 
+def embedding(input, weight, padding_idx=None,
+              max_norm=None, norm_type=2, scale_grad_by_freq=False,
+              sparse=False):
+    assert not sparse, 'sparse not supported in nn.functional.embedding for now'
```
Comment (albanD) [inline torch/nn/functional.py]: Why is this not supported?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -550,6 +550,19 @@ def bilinear(input1, input2, weight, bias=None):
...
+    assert padding_idx is None, 'non-None padding_idx not supported in nn.functional.embedding currently'
```
Comment (albanD) [inline torch/nn/functional.py]: Same as above, why is this not supported?

File: torch/nn/functional.py (hunk 1 above)
Comment (hughperkins) [inline torch/nn/functional.py]: because I dont know what it means, nor how to implement it.

Comment (hughperkins) [inline torch/nn/functional.py]: oh, `sparse`. ok, I know what sparse means, but not how to implemnet it :-)

Comment (albanD) [inline torch/nn/functional.py]: You can check how the [`nn.Module` version](https://github.com/pytorch/pytorch/blob/73fead9f8f4a53e64a078d3891ae43e865b32c7b/torch/nn/modules/sparse.py#L87-L94) works but I think this is completely handled by the underlying `Function`.

Comment (hughperkins) [inline torch/nn/functional.py]: Yeah, thats fair. Probably an artifact of when I wasnt sure how I would implement this.

Comment (hughperkins) [inline torch/nn/functional.py]: (updated)

File: torch/nn/functional.py (hunk 2 above)
Comment (hughperkins) [inline torch/nn/functional.py]: same reason. thoughts?

**Comment by hughperkins** [comment]
Question: is there anything outstanding I need to check for this?

**Comment by soumith** [comment]
add docs, dont raise error on padding_idx, have a test checking output against Module version

**Comment by hughperkins** [comment]
1. hows this looking?
2. how to run the doctests? (I tried `python torch/nn/functional.py` but didnt seem to work. I tried `grep -r doctest`, and found a conftest.py in sphinx, that contains that. So, maybe need to run some tests in sphinx somehow?

Note that I'm not sure that `padding_idx` makes sense in fucntional version? I think we just need to zero out the appropriate embedding_matrix rows ourselves? You can see how I think we should do that in the doctest bit:

```
        >>> # example with padding_idx
        >>> embedding_matrix = Variable(torch.rand(10, 3))
        >>> embedding_matrix[0].zero_()
        >>> input = Variable(torch.LongTensor([[0,2,0,5]]))
        >>> torch.nn.functional.embedding(input, embedding_matrix)

        Variable containing:
        (0 ,.,.) =
          0.0000  0.0000  0.0000
          0.3452  0.4937 -0.9361
          0.0000  0.0000  0.0000
          0.0706 -2.1962 -0.6276
        [torch.FloatTensor of size 1x4x3]
```
Thoughts?

**Comment by hughperkins** [comment]
Hi. Can you confirm any outstanding issues to be addressed on this please?

**Comment by hughperkins** [comment]
oh, testing against module output. fair enough...

**Review by soumith — APPROVED** [review]

**Comment by hughperkins** [comment]
Cool :-)

File: torch/nn/functional.py [hunk 3]
```diff
@@ -557,6 +557,73 @@ def bilinear(input1, input2, weight, bias=None):
...
+        Variable containing:
+        (0 ,.,.) =
+          0.0000  0.0000  0.0000
+          0.3452  0.4937 -0.9361
+          0.0000  0.0000  0.0000
+          0.0706 -2.1962 -0.6276
+        [torch.FloatTensor of size 1x4x3]
+
+    """
+    return torch.nn.backends.thnn.backend.Embedding(
```
Comment (fmassa) [inline torch/nn/functional.py]: Nit: it would be great if we could use the new-style interface here, as has already been updated in the [`module` interface](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/sparse.py#L91-L95). Also, is there a particular reason why `padding_idx` was left out of this PR?

Comment (hughperkins) [inline torch/nn/functional.py]: Well, I've never used padding_idx, but how were you imagining one would use padding_idx in the functional version?

Comment (fmassa) [inline torch/nn/functional.py]: I have never used the Embedding layer, so I don't really know the use-case. But it is exposed in the `module` interface, and ideally the `module` interface should use the `functional` interface, so in this case it would make sense to expose it as well.

Comment (hughperkins) [inline torch/nn/functional.py]: well, what I reckon padding_idx does is, it zeros out one row in our embedding matrix. However, in the functional version, we are providing the embedding matrix, so we can simply zero out that row ourselves, before passing it into the functional version?

Comment (fmassa) [inline torch/nn/functional.py]: The functional modification was already present in https://github.com/pytorch/pytorch/pull/1916, so disregard my comment on it. About `padding_idx`, I don't know much about it, but according to the comment in https://github.com/torch/nn/pull/650, it makes code faster in some cases. But I have no strong opinion on that

Comment (hughperkins) [inline torch/nn/functional.py]: oh, unless it's something like, there are two parts:
1. we create a blank row, that padding idx will use, and set padding idx to that, and then, missing this next bit at the moment :-P :
2. any indexes that go 'off the end' are replaced with padding idx, and then return blanks

Sounds fairly plausible perhaps?

Comment (apaszke) [inline torch/nn/functional.py]: IIRC `padding_idx` is faster, because it only zeros the rows that are actually used (it's lazy vs eager zero-fill)

Comment (hughperkins) [inline torch/nn/functional.py]: 'zeros the rows that are actually used'. Do you mean, in the output tensor?

Comment (apaszke) [inline torch/nn/functional.py]: Ok, I've just checked the C kernel, and it doesn't have any effect in forward, but it's a no-op in backward (gradient of padding is 0), so it can quickly skip the row
